man/ffpcol.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpcol 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpcol - This program creates columnar FFPs from key-value output of ffpry.
.SH SYNOPSIS
.BI "ffpcol [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
Given no options, the default behavior of the program is to
convert key-value FFPs to columnar format.  This utility is used to convert the output of
.BR "ffpry" ","
.BR "ffpaa" ","
or
.BR "ffptxt"
.RB " to the format required for " ffprwn " if the FFP" 
has been saved in key-value format.
FFPs will be read from standard input if no options are supplied and 
.B ffpcol  
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the file 
argument list.  The default input type is a key-value nucleotide FFP. 
When input is directed from a pipe a temporary file is created in the environmentally
defined TMP directory or /tmp if this variable is undefined.

.SH OPTIONS
.TP
.B \-h, --help
Display help command.
.TP
.B \-a, --amino
Input is an amino acid FFP. Default is nucleotide.
.TP
.B \-t, --text
Input is 26 letter text. Default is nucleotide.
.TP
.B \-d, --disable
Disable classing of either nucleotide or amino acid sequence.
.TP
.B \-V, --verbose
Be more verbose.
.TP
.B \-v, --version
Print version information
.TP
.B \-h, --help
This help message
.PP
.SH EXAMPLES
.PP
A columnar format file produced by 
.B ffpcol 
is the main interchange format to pass
FFPs to a number of other utilities.  Sample input from
a feature counter such as 
.B ffpry 
may resemble something like this:
.PP
.CODE ATG	2	TGC	 3	GGA	4 ...
.CODE TGC	1	CAC	 2	GGA	1 ...
.PP
From the key-value input format, 
.B ffpcol
will find all the features contained in all the FFPs
and align the feature frequency in the same tab delimited column,
printing out zeros if a feature is missing in a particular
row.  The actual sequence of the feature is discarded at
this point. The columnar format files produced from input like that above
will look like:
.PP
.CODE  2	3	0	4
.CODE  0	1	2	1
.PP
.RB "To convert a (key,value) FFP for use in " ffprwn, 
simply arrange the conversion process
as a pipeline, whereby the standard output of one utility is sent to the standard input of
another:
.PP
.CODE ffpry -l 2 *.fna | ffpcol | ffprwn
.PP
To disable classing use this format 
(take special node of the 
.B \-d
switch in both of the calls to
.B ffpry 
and 
.BR "ffpcol" "):"
.PP
.CODE ffpry -l 4 -d *.fna | ffpcol -d | ffprwn
.PP
For amino acids, use the 
.B \-a
option:
.PP
.CODE ffpaa -l 4 *.faa | ffpcol -a | ffprwn
.PP
To disable classing for amino acids use
the options
.B \-a
and 
.BR "\-d" ","
or in stacked form,
.BR "\-ad" ":"
.PP
.CODE ffpaa -l 4 -d *.faa | ffpcol -ad | ffprwn
.PP
Likewise for text input use
.BR "\-t" ":"
.PP
.CODE ffptxt -l 6 *.txt | ffpcol -t | ffprwn
.PP
Note when piping output from a utility into 
.B ffpcol
via a pipe that a temp file is created ( from the output of
preceding command ), whereas input from file 
arguments or via input redirection (i.e. 
.CW ffpcol < file 
) doesn't require temp files.  Temporary files are written
to 
.CW /tmp
-- upon creation they are immediately unlinked
so they are effectively 'hidden' temp files.
.SH FUTURE DIRECTIONS
Conversion to Rabin-Karp Hash function.
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpry(1),
.BR ffpaa(1)
.BR ffptxt(1)
.BR ffprwn(1),
.BR ffpjsd(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpcol.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "hash.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "../config.h"

#define MAX_WORD_LENGTH 40 /**< Maximum feature length */

char PROG_NAME[FILENAME_MAX];

void hashCol(FILE * fp);
int isKeyBased(FILE * fp);
int getKeyLength(FILE * fp);

char *weightVector;
int Length = MAX_WORD_LENGTH;
			    /**< Feature length */



char usage_str[] = "Usage: %s [OPTIONS]... [FILE]...\n\
This program creates columnar FFPs from key val output of ffpry,\n\
ffpaa, or ffptxt.\n\n\
Given no options, the defeault behavior of the program is to\n\
generate columnar FFPs for Nucleic acids.\n\
\t-a, --amino\tInput is Amino acid\n\
\t-t, --text\tInput is text\n\
\t-d, --disable\tDisable classing of AAs and Nuc.\n\
\t-V, --verbose\tBe more verbose.\n\
\t-h, --help\tThis text.\n\
\t-v, --version\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


bool flagV = false;


int main(int argc, char **argv)
{
    FILE *fp = NULL;
    int opt;
    bool flagA = false;
    bool flagT = false;
    bool flagD = false;

    int mode = nucleotide;
    int option_index = 0;

    static struct option long_options[] = {
	{"help", no_argument, 0, 'h'},
	{"amino", no_argument, 0, 'a'},
	{"text", no_argument, 0, 't'},
	{"disable", no_argument, 0, 'd'},
	{"version", no_argument, 0, 'v'},
	{"verbose", no_argument, 0, 'V'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();
    
    strcpy( PROG_NAME, basename(argv[0]) );

    while ((opt = getopt_long(argc, argv, "hatdvV",
			      long_options, &option_index)) != -1)
	switch (opt) {
	case 'a':
	    flagA = !flagA;
	    break;
	case 't':
	    flagT = !flagT;
	    break;
	case 'd':
	    flagD = !flagD;
	    break;
	case 'V':
	    flagV = !flagV;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    break;
	}


    if (flagA && flagT)
	fatal_msg("Option -a or -t not both\n");


    if (flagA)
	mode = amino;
    else if (flagT)
	mode = text;
    else {
    }

    initHash(0, mode, !flagD);


    if ((argc - optind) > 1)
	fatal_msg("Specify only one file argument.\n");

    argv += optind;

   do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL){
		    fatal_msg("%s: %s.\n", *argv,strerror(errno));
		
	    }

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO)) {
		printErrorUsageStr();
	}

	if (!isRegularFile(fp)) 
	    fp = convertPipeToFile(fp);

        if (!isKeyBased(fp))
	    fatal_msg("%s: Not a key valued FFP.\n", *argv);




	Length = getKeyLength(fp);
	hashCol(fp);

	fclose(fp);

    } while (*argv);

    return EXIT_SUCCESS;
}


/**
 * Convert a (key,value) FFP to a columnar FFP
 *
 * FFP in fp must be columnar FFP.
 *
 * @param fp A file pointer to a (key,value) FFP
 * @return none
 */



void hashCol(FILE * fp)
{
    //for accepting input via pipe      
    // Used in all cases
    unsigned val;
    char * s;
    char **keys;
    unsigned i;
    char c[3];
    int items;
    unsigned lineno=1;
    off_t line_offset=0;



    s=(char *) chkmalloc(sizeof(char),(Length+1));

//@TODO move this comment to description of convertFilePtr

// cut here
// If file is stdin we need to create a tmp file
// This is equivalent to to:
//  stdin | cat > /tmp/tmp.1a321asf 
// However we don't want a file passed to the
// program via  ffpcol < infile, to undergo this
// process, since a file opened in this manner
// is seekable and doesn't need to be copied.
// File opened this way ffpcol <(ffpry *.fna) 
// i.e. via process substitution is not seekable

    // Find all keys in the file.

    errno=0;

   

    //@TODO see if this can be sped up by using %*u to skip reading and conversion of val.
    while ( (items=fscanf(fp, "%s %u%[\r\n]", s, &val,c)) != EOF ) { 
	//if (errno) 
	//	fatal_msg("Parse error at line %u char %ld: %s. %d\n",
	//		lineno,ftell(fp)-line_offset, strerror(errno),errno );
	
	switch (items) {
	        case 3:
			if (flagV)
				fprintf(stderr,"Processed Line: %d\n",lineno);
			line_offset=ftell(fp);
			lineno++;	  
		case 2:
			hashInc(s);		// increments hash by value;
		break;	
		default:
			fatal_msg("Parse error at line %u char %ld.\n",
			lineno,ftell(fp)-line_offset);
		break;
	}
    }

    if (ferror(fp)) 
    	fatal_msg("Read Error: %s",strerror(errno));

    hashKeys(&keys);

    rewind(fp);

    fflush(fp); // Flush last character read


    //zero out the hash

    for (i = 0; i < numKeys(); i++) {
	hashAssign(keys[i], 0);
    }


	while ((items = fscanf(fp, "%s %u%[\r\n]", s, &val, c)) != EOF) {

	    switch (items) {
		    case 2:
	    		hashAssign(s, val);
		    break;
		    case 3:
	    		hashAssign(s, val);
	   		for (i = 0; i < numKeys()-1; i++) {
				printf("%u\t", hashval(keys[i]));
				hashAssign(keys[i], 0);
	   		}
	    		printf("%u\n",hashval(keys[i]));
	    		hashAssign(keys[i], 0);
		    break;
	    }
	}

  free(s);
  // No need to free keys, since program terminates
}


